### ISSUE TITLE
Generator return types via __iter__ not applied to yield-from expressions

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#1676]
(bringing https://github.com/microsoft/pyright/issues/1652#issuecomment-802734763 into its own issue)

### Describe the bug

When implementing `__iter__()` to provide a generator with a return type (rather than the class being an iterator itself), the value of a `yield from` expression is of an unknown type.

### To reproduce

```python
from typing import Generator

class Yielder:
    def __iter__(self) -> Generator[int, None, bool]:
        yield 1
        return True

def collect():
    y = Yielder()
    ok = yield from y  # unknown type
    return ok
```

Fetching the generator directly with `yield from y.__iter__()` gains the `bool` type as expected.

### Expected behavior

It would be useful for callers who know the full type signature to be able to make use of it.

At a guess, the iterator becomes an `Iterator[int]` somewhere as part of the `__iter__` protocol.

### VS Code extension or command-line

Pyright 1.1.124 in VSCode.

### Additional context

The default Python extension in VS Code identifies the result itself as a generator (usually it would be `NoneType`), which suggests it's doing _something_ with generator return types, if not the right thing:

![image](https://user-images.githubusercontent.com/4025899/112115993-b8277d00-8bb1-11eb-9028-0671ee4d624b.png)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.125, which I just published. It will also be included in the next release of pylance.
